### PR TITLE
Refactor layer slice/dims/view/render state

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description

This is a pure refactor of some of the code associated with slicing a layer. There should be no public API or behavior change. It groups a few `Layer` attributes (`_ndisplay`, `_dims_point`, `_dims_order`) into a (terribly named) deeply immutable dataclass `SliceInput`. It also removes some existing `Layer` properties derived from those attributes instead favoring access through the new attribute/class or avoiding some confusing call chains entirely.

There are a few motivations for this PR.

- We have a long term goal to remove slice state from the layer. Grouping some of that state should make it easier to remove in the future.
- Grouping the slice state makes it easier to see what depends on it (answer: a lot!).
- It helps many possible implementations of async slicing. By grouping some of the slice state together, it makes it much easier to guard read/write access to that state (e.g. with a lock), which is critical if multiple threads are reading/writing it.
- Some of existing slicing code (particularly around dimension indices) is confusing and this PR helps simplify it a little.

I have mixed feelings about this PR, but think some of the simplifications and new comments/docs are useful and should get merged in some form. In particular, I am seeking a better name than `SliceInput`, so please don't be shy with new name suggestions!

## Type of change
- [x] Refactor

# References
Related to #4795 

# How has this been tested?
- [ ] all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/_qt/qt_viewer.py
Comment (andy-sweet) [inline napari/_qt/qt_viewer.py:1096]: I believe this should always be `True` as the number of dimensions in the viewer should always be greater than equal to the number of dimensions in any one layer, but if someone knows better, please correct me and I'll put this back in.

File: napari/layers/base/base.py
Comment (andy-sweet) [inline napari/layers/base/base.py:1574]: This chain of function/property calls was particularly fun. I'm pretty sure this simplifies to just getting the displayed dims (because the call to `displayed_axes` first sorts then unsorts the displayed dimensions), but feel free to correct.

File: napari/layers/base/base.py
Comment (andy-sweet) [inline napari/layers/base/base.py:647]: Removing `_ndisplay`, `_dims_displayed`, and `_dims_not_displayed` causes a lot of the deltas in this PR. I'd be willing to bring these back to reduce the size of this PR and make references to these things slightly more concise.

File: napari/layers/base/base.py
Comment (andy-sweet) [inline napari/layers/base/base.py:1281]: I think we expect `offset` to always be non-negative (because the number of dimensions in the world should always be greater than equal to the number of dimensions in any one layer), so I wonder if we could just remove this condition. But more informed opinions are welcome!

File: napari/layers/base/base.py
Comment (andy-sweet) [inline napari/layers/base/base.py:1549]: I believe this was unused.

**Comment by andy-sweet** [comment]
@alisterburt : I assigned you to review because you had some questions when I mentioned this PR in a recent community meeting.

@jni : I assigned you to review because we talked about this in a recent architecture working group.

Other reviews welcome!

File: napari/layers/image/experimental/octree_image.py
Comment (andy-sweet) [inline napari/layers/image/experimental/octree_image.py:364]: This used the sorted displayed dimensions, whereas the equivalent code in `Layer` uses the unsorted ones. I'm not familiar enough with the experimental async implementation to know if that's intended or not. But there should be no functional change from main here.

File: napari/layers/base/base.py
Comment (andy-sweet) [inline napari/layers/base/base.py:1253]: I just noticed the following similar function: https://github.com/napari/napari/blob/87a263328fd71c8d3fc6252fe8ee2d09056ae776/napari/layers/utils/layer_utils.py#L657

Maybe we can just use one for this correspondence? Or there a difference I'm not seeing?

File: napari/layers/base/base.py
Comment (alisterburt) [reply napari/layers/base/base.py:647]: I don't think this is necessary, this will be a nice clean break

edit: this was a reply to the suggestion of not removing these properties. I am pro removing these properties

File: napari/layers/base/base.py
Comment (alisterburt) [reply napari/layers/base/base.py:1253]: I think the difference is that the former accounts for the full set of dims, the latter only the displayed dims. IIRC there are some subtleties (i.e. probably illogical slicing behaviour) which means these should be dealt with very carefully - @kevinyamauchi may have some more specific info here

File: napari/layers/base/base.py
Comment (alisterburt) [reply napari/layers/base/base.py:1281]: I agree with your expectation here and don't know why this would ever be negative, possibly some strange initialisation?

File: napari/layers/points/_points_mouse_bindings.py
Comment (alisterburt) [inline napari/layers/points/_points_mouse_bindings.py:172]: c.f. discussion above, I'm not 100% that the method which handles slicing down of world dims handles the displayed dims properly, would you mind checking that this change is covered by tests before merge?

File: napari/layers/points/_tests/test_points_mouse_bindings.py
Comment (alisterburt) [inline napari/layers/points/_tests/test_points_mouse_bindings.py:79]: clean!

File: napari/layers/utils/_slice_input.py
Comment (alisterburt) [inline napari/layers/utils/_slice_input.py]: these descriptions are 😍

File: napari/layers/utils/_slice_input.py
Comment (alisterburt) [inline napari/layers/utils/_slice_input.py:61]: am I right in saying that this becomes the 'single source of truth' for how to slice data when you have mixed dimensionality? if so I love it.

I don't love the name here - what do we think about `reduce/lower_dimensionality` as an alternative?

File: napari/layers/utils/_slice_input.py
Comment (alisterburt) [inline napari/layers/utils/_slice_input.py:15]: I'm wondering about dataclass vs pydantic model here - would a model with `allow_mutation=False` in the config make more sense in terms of consistency with other parts of the codebase?

**Review by alisterburt — COMMENTED** [review]
This is a beautiful piece of work Andy, really tidies some things up! 

I have one or two little questions I would like to hear responses to before approval but otherwise I think this is good to go. 

Re: the name, `SliceInput` is okay, slight preference for `SlicingInput` although it's longer.

I feel like the `_SliceInput` class is so fundamental it should probably be a component, what do you think about that idea?

File: napari/layers/utils/_slice_input.py
Comment (andy-sweet) [reply napari/layers/utils/_slice_input.py:61]: No. This is really just for `Layer._update_dims`, and only does anything interesting when the `ndim` changes. We call `_update_dims` many times, though I think `ndim` will only change when the `data` setter with data of new dimensionality (which isn't something I would expect to happen much, if ever). In general, I'd love to replace lots of our usage of `_update_dims` with something like `refresh` and/or `_update_extent`.

The dimension correspondence problem is still handled by other functions/methods including `reorder_after_dim_reduction` here (which I'm not convinced is always doing something meaningful). Ideally, we would have one function to handle dimension correspondence. My sense is that we have a few things that mostly work in practice (but we know doesn't in others), but I'm still trying to understand all our usage and intended behavior there. Ultimately, having labeled dimensions would make this much simpler to understand, but we could probably do with a mostly working simple solution before then.

**Comment by alisterburt** [comment]
marking this as a good one to get merged for 0.4.17 - please ping for re-review as and when/let me know when you think this is ready to merge

**Comment by Carreau** [comment]
The only real failure I seem to see is a circular import using pytest (edited for readability):

```
napari_builtins/_tests/conftest.py:10: in <module>
    from napari import layers
napari/layers/__init__.py:11: in <module>
    from .base import Layer
napari/layers/base/__init__.py:1: in <module>
    from .base import Layer, no_op
napari/layers/base/base.py:35: in <module>
    from ..utils._slice_input import _SliceInput
napari/layers/utils/_slice_input.py:9: in <module>
    from napari.components.dims import reorder_after_dim_reduction
napari/components/__init__.py:19: in <module>
    from .layerlist import LayerList
napari/components/layerlist.py:9: in <module>
    from ..layers import Layer
E   ImportError: cannot import name 'Layer' from partially initialized module 'napari.layers' (most likely due to a circular import) (napari/layers/__init__.py)



layer.base -> layer.base.base -> utils._slice_input -> napari.component.dims:reorder_after_dim_reduction -> layerlist:LayerList -> layer.Layer
```

**Comment by Carreau** [comment]
it looks like we can move `reorder_after_dim_reduction` somewhere else (utils ?), and that should avoid circular imports

File: napari/layers/base/base.py
Comment (brisvag) [reply napari/layers/base/base.py:647]: I agree with @alisterburt, keeping them around is more confusing than not.

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py]: Is the use of assert here intended for performance improvement from debug to production? If not, we should use normal `raise` syntax to avoid missing the error.

File: napari/layers/base/base.py
Comment (brisvag) [reply napari/layers/base/base.py:1281]: Could this come into play when a new layer with more dims than thew world gets added to the viewer?

**Review by brisvag — COMMENTED** [review]
Looks great, amazing work @andy-sweet! I've only looked through `layers.base` for now, but so far so good. I'll get back to it tomorrow.

**Comment by andy-sweet** [comment]
Thanks for all the comments so far. I'll respond properly tomorrow.

> marking this as a good one to get merged for 0.4.17 - please ping for re-review as and when/let me know when you think this is ready to merge

As there's no intended behavior change / bug fixing here (i.e. this is supposed to be a pure refactor), I'd suggest this is better for 0.5 in the spirit of only bringing in what's needed for 0.4.17.

File: napari/layers/utils/_slice_input.py
Comment (kephale) [reply napari/layers/utils/_slice_input.py:15]: There are some differences between their behavior: https://stackoverflow.com/questions/62011741/pydantic-dataclass-vs-basemodel

It looks like type enforcement on the init function for `dataclass` does not happen immediately. This would impact the way `_SliceInput` is consistently used in this PR. However, `dataclass` seems to outperform the pydantic `BaseModel` (https://stefan.sofa-rockers.org/2020/05/29/attrs-dataclasses-pydantic/ and https://jackmckew.dev/dataclasses-vs-attrs-vs-pydantic.html). These performance results would suggest considering attrs (https://github.com/python-attrs/attrs) instead of `dataclass` or pydantic.

Since the class is immutable, the cost of validation is only incurred at init. It might be worth it to stick to pydantic for consistency *and* to have more validation. If performance becomes an issue, then reconsider.

File: napari/layers/base/base.py
Comment (andy-sweet) [reply napari/layers/base/base.py:1281]: My assumption is that is intended to never happen (because `Dims` should be updated first) and the order of the code right now in `ViewerModel._on_add_layer` means that it doesn't. I guess it's possible for a layer that hasn't been added yet to the viewer to trigger this function, though I'm unsure of any public ways to do that. Again, I assume the intent is that calling this only makes sense for layers that have been added to the viewer given that it is finding correspondences between the layer's dimensions and those displayed (presumably in the viewer).

**Comment by andy-sweet** [comment]
> it looks like we can move `reorder_after_dim_reduction` somewhere else (utils ?), and that should avoid circular imports

It looks like you removed the function definition, but didn't actually add the new module. I actually just used `numpy.argsort` instead in `_slice_input.py`, as there should only be 2 or 3 values to sort and this will happen very rarely. I may actually remove that utility function in 0.5 as well, so didn't really want to add a module just for it.

**Review by kephale — APPROVED** [review]

File: napari/layers/utils/_slice_input.py
Comment (andy-sweet) [reply napari/layers/utils/_slice_input.py:15]: My reason to use the built-in `dataclass` over pydantic's `BaseModel` here is that it's enough. `_SliceInput` is intended to be a private class that groups together a few attributes that we want to reason about together. We don't need validation (because the input is not coming from the user) or (de)serialization, so there doesn't seem to be a strong reason to jump to `BaseModel`.

Even though performance isn't a big motivator here, pydantic is about 20x slower than the built-in dataclass to create this object for 3D display.

```python
In [1]: from dataclasses import dataclass

In [2]: @dataclass
   ...: class _SliceInput:
   ...:     ndisplay: int
   ...:     point: tuple[float, ...]
   ...:     order: tuple[int, ...]
   ...: 

In [3]: %timeit _SliceInput(ndisplay=3, point=(0, 0, 0), order=(3, 0, 1))
399 ns ± 3.54 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: from pydantic import BaseModel

In [5]: class _SliceInput(BaseModel):
   ...:     ndisplay: int
   ...:     point: tuple[float, ...]
   ...:     order: tuple[int, ...]
   ...: 

In [6]: %timeit _SliceInput(ndisplay=3, point=(0, 0, 0), order=(3, 0, 1))
10.6 µs ± 116 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

10 microseconds isn't much, but that is on a fairly modern laptop and these objects may be created fairly frequently (e.g. imagine moving the slider or panning/zooming around the canvas), so it's best to keep this as light as possible.

I kind of understand the argument of consistency, so I would be willing to bend here, but will likely stick with `dataclass` unless forced to change it.

Comment (andy-sweet) [reply napari/layers/utils/_slice_input.py:15]: For completeness, I did also try `attrs`, which is maybe a little faster than the built-in `dataclass`.

```python
In [1]: from attrs import define

In [2]: @define
   ...: class _SliceInput:
   ...:     ndisplay: int
   ...:     point: tuple[float, ...]
   ...:     order: tuple[int, ...]
   ...: 

In [3]: %timeit _SliceInput(ndisplay=3, point=(0, 0, 0), order=(3, 0, 1))
368 ns ± 32.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

But it's hard to justify it as a new explicit dependency based on that. It would also create more inconsistency in terms of when we should use `dataclasses`, `attrs`, and `pydantic`.

File: napari/layers/base/base.py
Comment (andy-sweet) [reply napari/layers/base/base.py]: I mostly added it to quickly check if that case was ever being hit mostly to check if we ever hit in the `offset < 0` case in `_world_to_data_dims_displayed`. I'll probably just remove it for now, as I think we should be able to progress in that case (as we do on `main` now).

File: napari/layers/utils/_slice_input.py
Comment (kephale) [reply napari/layers/utils/_slice_input.py:15]: Thank you! Totally agree on all points. 

Nice to see that the timing results align with the reported comparisons between `dataclass`, `attrs`, and `pydantic.BaseModel`. From what I see about `attrs` I might choose it instead of `pydantic` when making an initial dependency selection: get the speed of `dataclass` plus ability to turn on validations later down the road.

File: napari/layers/utils/_slice_input.py
Comment (liu-ziyang) [inline napari/layers/utils/_slice_input.py:23]: I don't have a strong opinion here because I don't know if the naming is following some convention. But our current naming is very generic and requires docs to note their semantics. Some variable renaming would be sweet

Comment (andy-sweet) [reply napari/layers/utils/_slice_input.py:23]: The naming is mostly the existing naming. I'm unsure if it's following a convention outside of napari. Happy to take suggestions, but my main focus was to at least clarify what these things are with docstrings. If there are ways to make concise versions of those that can be directly reflected in the names, that would be great.

**Comment by andy-sweet** [comment]
Would be great to be able to merge this soon as we start work on implementing async slicing.

@brisvag , @alisterburt: as you've already taken a look (a while ago, mind), would be great to get an approval if everything looks good. Also happy to address issues or break this big PR up.

File: napari/components/dims.py
Comment (brisvag) [inline napari/components/dims.py]: Codecov marks this as not covered, maybe we can add a small test since we're here?

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py:665]: This simplification feels great :)

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py:1265]: EDIT: the more I think about this, and the less I feel I understand it... I leave my comment below, but I'm no longer convinced I got it :/

---

Ok, so I got quite confused and double backed a couple times when reading this docstring. I think I understand now, so I propose to add a clearer schematic to the dosctring for future reference. What I understand from the above is basically this (which is WRONG):

```
[2, 1, 0, 3]
    |  |  |
    | [0, 1]
    |  |  |
   [1, 0, 2]
```

However, this makes no sense. Instead, this function returns the *indices* that I can use to index into the layer data corresponding to the world dimensions.

So the right way to look at this is as follows: we right-align dimensions like this:

```
[0, 1, 2, 3]
    |  |  |
    | [0, 1]
    |  |  |
   [0, 1, 2]
```

Then we change the order:

```
[2, 1, 0, 3]
 |  |     |
[0, |     1]
 |  |     |
[1, 0,    2]
```

And we get the final pairings. Sorry if this was obvious to you, but it really confused me :P

What I'm *still* confused about it the role of `ndim_world`. Isn't that just `len(world_dims)`?

File: napari/layers/image/image.py
Comment (brisvag) [inline napari/layers/image/image.py]: There were several places with this code + comment 
```
# equivalent to: order = np.argsort(displayed)
order = sorted(range(len(displayed)), key=lambda x: displayed[x])
````

I think it was for performance reasons. Is it worth considering here as well?

File: napari/layers/shapes/_shapes_models/shape.py
Comment (brisvag) [inline napari/layers/shapes/_shapes_models/shape.py]: Same as above (`argsort` vs `sorted(range...`)

File: napari/layers/utils/_slice_input.py
Comment (brisvag) [reply napari/layers/utils/_slice_input.py:15]: > (e.g. imagine moving the slider or panning/zooming around the canvas),

This shouldn't happen on panning/zooming, only on dims changes. But yeah, I agree that dataclass is completely fine here.

File: napari/layers/utils/_slice_input.py
Comment (brisvag) [reply napari/layers/utils/_slice_input.py:23]: For this refactor, I suggest to keep these names (so the rest of the diff looks as close as possible to how it was before), but I agree that we can make these better.

File: napari/layers/utils/_slice_input.py
Comment (brisvag) [inline napari/layers/utils/_slice_input.py]: Same `argsort` vs `sorted(range` comment as above.

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py:660]: I'm not very familiar with the transform system, but I noticed you left this stateful update outside of the `SliceInput`. Is that inevitable? Why do transforms themselves need to change when slicing changes?

**Review by brisvag — CHANGES_REQUESTED** [review]
Ok, most of this looks good to me. I followed step-by-step through the conversion from the old `_slice_dims` nonsense to the new methods on `SliceInput`, and they all make sense to me: great job @andy-sweet!

I left a few comments here and there; the main issue for me is the `_world_to_layer_dims` shenanigans: I'm not convinced that this and `dims_displayed_world_to_layer` are equivalent... And to be fair, I'm not convinced the function does what it's intended to do. I hope @kevinyamauchi and @alisterburt can be more helpful than me here in reviewing this part!

File: napari/layers/image/image.py
Comment (andy-sweet) [reply napari/layers/image/image.py]: Maybe, but keep in mind we're always sorting 2 or 3 numbers. I can write some really fast sorting algorithms for that case :p

File: napari/layers/base/base.py
Comment (andy-sweet) [reply napari/layers/base/base.py:660]: Great question. When we do slicing `_update_dims` is called, but neither of these conditions will be true, so the transforms won't actually be modified.

I have it on my bucketlist to clean up / reduce use of `_update_dims`, but not in this PR. I think it's only really needed when `data` is set, since the new `data` could actually have new dimensionality. We also call it in the transform property setters, which is to clear our the extent cache, but again dimensionality should not actually change.

File: napari/layers/base/base.py
Comment (andy-sweet) [reply napari/layers/base/base.py:1265]: Don't worry. I'm still confused too. I tried to improve things with some renaming and docstrings, but looks like it didn't help too much :D

I think it's a good goal to of this PR to clarify what is going on now, though I think fixing and simplify everything is out of scope.

Agreed that some better schematics here help. I usually add names to the dimensions of each layer (e.g. `a0`, `b2`), so that I can keep track of the correspondences while being too lazy to maintain ascii art. I'll try adding something in this PR though.

As for `ndim_world`, I'm pretty sure it is when this is called in `Layer._slice_dims`, but I'm not sure if that's also true for `_points_mouse_bindings._update_drag_vectors_from_event`. Will check and see if we can simplify that.

File: napari/layers/base/base.py
Comment (brisvag) [reply napari/layers/base/base.py:660]: I see; Is it poossible to generalize transforms so that if you combine a 3D affine with a 4D affine, everything works? This way we wouldn't have to wirry about this at the layer level...

Comment (andy-sweet) [reply napari/layers/base/base.py:660]: > I see; Is it poossible to generalize transforms so that if you combine a 3D affine with a 4D affine, everything works? This way we wouldn't have to wirry about this at the layer level...

Maybe - depends what you mean by everything works. Personally, I want to be able to explicitly assign dimensions to the layer's transform. If someone passes a transform that does not explicitly account for some dimension, we can have some logic to handle that, but in some cases should just error. In which case, I prefer to raise early and handle things at the layer level. Otherwise, we'll error when trying to make correspondences between layers which seems like it would be confusing.

Remember, this section isn't really about slicing. It's about `data` changing.

File: napari/layers/shapes/_shapes_models/shape.py
Comment (andy-sweet) [reply napari/layers/shapes/_shapes_models/shape.py]: I turned our simple `argsort` in `utils.misc` into a public function, documented it, and used it here to avoid a performance regression.

File: napari/components/dims.py
Comment (andy-sweet) [inline napari/components/dims.py]: @brisvag : I think you made a comment on the lack of test coverage for this property, but it got wiped due to codecov rerunning. The lack of coverage is probably because we don't use it anywhere internally in napari. So I added to an existing test in `test_dims` to cover these three related properties.

File: napari/layers/base/base.py
Comment (andy-sweet) [reply napari/layers/base/base.py:1265]: I think we're basically repeating the conversation in #3882.

My understanding is that the interpretation you got from the current docstring is correct in the sense that this is what napari does now. As @jni put it

> the mapping of dimensions between different arrays changes depending on the dims being displayed!!!

Another way to think about it is that napari is trying to show as many dimensions of each layer as possible.  I added a test to show this.

As discussed in #3882, what napari does now may not be what we want going forward. What you described (right aligning then reordering) is one of the suggestions. One problem with that approach is that the order of `[2, 1, 0, 3]` and `ndisplay` of 2 implies that we're visualizing dimensions 0 and 3 of the first layer, but 0 doesn't have a corresponding dimension in the other two layers, which means that these will be 1D slices and we're unsure if we should show nothing or that row broadcasted.

This PR is definitely not the place to change the current behavior. But I have added the test above (so we know when it changes).

File: napari/layers/base/base.py
Comment (andy-sweet) [reply napari/layers/base/base.py:1253]: I'm going to leave both for now, but am hoping we can simplify the dimension correspondence logic soon.

File: napari/layers/points/_points_mouse_bindings.py
Comment (andy-sweet) [reply napari/layers/points/_points_mouse_bindings.py:172]: There's no functional change here. This is just a name change. Agreed there may be dimensional correspondence bugs out there, but I don't think this hefty PR is the place to fix those. I did add some tests to cover `_world_to_layer_dims` in `test_viewer_model.py` to express some of our current logic.

**Comment by andy-sweet** [comment]
I think I addressed enough here to get this in.

@alisterburt  and @brisvag, please take another look when you can.

File: napari/layers/base/base.py
Comment (brisvag) [reply napari/layers/base/base.py:1265]: Ok, I think I got it now, thanks! I guess the example above it just *happened* to work out the same way with the two different approaches ^^'

**Review by brisvag — APPROVED** [review]
Looks good to me!

**Comment by andy-sweet** [comment]
Thanks for the update - will merge this after 24 hours.

**Comment by kephale** [comment]
Woohoo, and I might have another use for these changes during the makeathon.

**Comment by andy-sweet** [comment]
Marked this with the v0.5.0 milestone, but given how much this touches, you might need to bring it in for 0.4.18

**Comment by andy-sweet** [comment]
Moved to 0.4.18 because this is large and predates most of the 0.4.18 PRs.

**Pull request merged**
